src/ray/object_manager/pull_manager.cc

```diff
@@ -39,33 +42,183 @@ uint64_t PullManager::Pull(const std::vector<rpc::ObjectReference> &object_ref_b
     it->second.bundle_request_ids.insert(bundle_it->first);
   }
 
+  // We have a new request. Activate the new request, if the
+  // current available memory allows it.
+  UpdatePullsBasedOnAvailableMemory(num_bytes_available_);
+
   return bundle_it->first;
 }
 
+bool PullManager::ActivateNextPullBundleRequest(
+    std::map<uint64_t, std::vector<rpc::ObjectReference>>::iterator next_request_it,
+    std::unordered_set<ObjectID> *object_ids_to_pull) {
+  // Check that we have sizes for all of the objects in the bundle. If not, we
+  // should not activate the bundle, since it may put us over the available
+  // capacity.
+  for (const auto &ref : next_request_it->second) {
+    auto obj_id = ObjectRefToId(ref);
+    const auto it = object_pull_requests_.find(obj_id);
+    RAY_CHECK(it != object_pull_requests_.end());
+    if (!it->second.object_size_set) {
+      RAY_LOG(DEBUG) << "No size for " << obj_id << ", canceling activation for pull "
+                     << next_request_it->first;
+      return false;
+    }
+  }
+
+  // Activate the bundle.
+  for (const auto &ref : next_request_it->second) {
+    auto obj_id = ObjectRefToId(ref);
+    bool start_pull = active_object_pull_requests_.count(obj_id) == 0;
+    active_object_pull_requests_[obj_id].insert(next_request_it->first);
+    if (start_pull) {
+      RAY_LOG(DEBUG) << "Activating pull for object " << obj_id;
+      // This is the first bundle request in the queue to require this object.
+      // Add the size to the number of bytes being pulled.
+      // NOTE(swang): The size could be 0 if we haven't received size
+      // information yet. If we receive the size later on, we will update the
+      // total bytes being pulled then.
+      auto it = object_pull_requests_.find(obj_id);
+      RAY_CHECK(it != object_pull_requests_.end());
+      num_bytes_being_pulled_ += it->second.object_size;
+
+      object_ids_to_pull->insert(obj_id);
+    }
+  }
+
+  // Update the pointer to the last pull request that we are actively pulling.
+  RAY_CHECK(next_request_it->first > highest_req_id_being_pulled_);
+  highest_req_id_being_pulled_ = next_request_it->first;
+  return true;
+}
+
+void PullManager::DeactivatePullBundleRequest(
+    std::map<uint64_t, std::vector<rpc::ObjectReference>>::iterator request_it,
+    std::unordered_set<ObjectID> *object_ids_to_cancel) {
+  for (const auto &ref : request_it->second) {
+    auto obj_id = ObjectRefToId(ref);
+    RAY_CHECK(active_object_pull_requests_[obj_id].erase(request_it->first));
+    if (active_object_pull_requests_[obj_id].empty()) {
+      RAY_LOG(DEBUG) << "Deactivating pull for object " << obj_id;
+      auto it = object_pull_requests_.find(obj_id);
+      RAY_CHECK(it != object_pull_requests_.end());
+      num_bytes_being_pulled_ -= it->second.object_size;
+      active_object_pull_requests_.erase(obj_id);
+      if (object_ids_to_cancel) {
```

Comment:
Why not just `RAY_CHECK(object_ids_to_cancell != nullptr)`?

Reply:
We don't always need to know. I only use it to prevent unnecessary object pulls when updating, since sometimes this will activate then deactivate a request that is right at the boundary of the current capacity.